src/D2L.Bmx/UpdateHandler.cs

```diff
@@ -0,0 +1,130 @@
+			File.Move( currentProcessPath, backupPath );
+		} else {
+			throw new Exception( "Could not get current process path" );
+		}
+
+		try {
+			var archiveRes = await httpClient.GetAsync( downloadUrl, HttpCompletionOption.ResponseHeadersRead );
+			using( var fs = new FileStream( downloadPath, FileMode.Create, FileAccess.Write, FileShare.None ) ) {
+				await archiveRes.Content.CopyToAsync( fs );
+				await fs.FlushAsync();
+				fs.Dispose();
+			}
+		} catch( Exception ex ) {
+			File.Move( backupPath, currentProcessPath );
+			throw new BmxException( "Failed to download the update", ex );
+		}
+
+		try {
+			string extension = Path.GetExtension( downloadUrl ).ToLowerInvariant();
+			string? extractPath = Path.GetDirectoryName( currentProcessPath );
+
+			if( extension.Equals( ".zip" ) ) {
+				DecompressZipFile( downloadPath, extractPath! );
+			} else if( extension.Equals( ".gz" ) ) {
+				DecompressTarGzipFile( downloadPath, extractPath! );
+			} else {
+				throw new Exception( "Unknown archive type" );
+			}
+			string newExecutablePath = Path.Combine(
+				extractPath!,
+				Path.GetFileName( currentProcessPath )!
+			);
+		} catch( Exception ex ) {
+			File.Move( backupPath, currentProcessPath );
+			throw new BmxException( "Failed to update with new files", ex );
+		} finally {
+			File.Delete( downloadPath );
+		}
+	}
+
+	private static string GetOSFileName() {
+
+		if( RuntimeInformation.IsOSPlatform( OSPlatform.Windows ) ) {
+			return "bmx-win-x64.zip";
+		} else if( RuntimeInformation.IsOSPlatform( OSPlatform.OSX ) ) {
+			return "bmx-osx-x64.tar.gz";
+		} else if( RuntimeInformation.IsOSPlatform( OSPlatform.Linux ) ) {
+			return "bmx-linux-x64.tar.gz";
+		} else {
+			throw new Exception( "Unknown OS" );
+		}
+	}
+
+	public static void Cleanup() {
+		string processDirectory = Path.GetDirectoryName( Environment.ProcessPath ) ?? string.Empty;
+		if( string.IsNullOrEmpty( processDirectory ) ) {
+			return;
+		}
+		foreach( string file in Directory.GetFiles( processDirectory, "*.old.bak" ) ) {
+			try {
+				File.Delete( file );
+			} catch( Exception ex ) {
+				Console.Error.WriteLine( $"WARNING: Failed to delete old version {file}" );
+			}
+		}
+	}
+
+	private static void DecompressTarGzipFile( string compressedFilePath, string decompressedFilePath ) {
+		string tarPath = Path.Combine( decompressedFilePath, "bmx.tar" );
+		using( FileStream compressedFileStream = File.Open(
+			compressedFilePath,
+			FileMode.Open,
+			FileAccess.Read,
+			FileShare.Read )
+		) {
+			using FileStream outputFileStream = File.Create( tarPath );
+			using var decompressor = new GZipStream( compressedFileStream, CompressionMode.Decompress );
+			decompressor.CopyTo( outputFileStream );
+		}
+
+		try {
+			TarFile.ExtractToDirectory( tarPath, decompressedFilePath, true );
+		} finally {
+			File.Delete( tarPath );
+		}
+	}
+
+	private static void DecompressZipFile( string compressedFilePath, string decompressedFilePath ) {
```

Comment:
```suggestion
	private static void ExtractZipFile( string compressedFilePath, string decompressedFilePath ) {
```

Wonder if `Extract` makes more sense here